server/src/instant/db/instaql.clj

```diff
@@ -1967,53 +1977,163 @@
+                                   :where-clauses (:where-clauses result)})
+                 (catch Exception e
+                   (tracer/with-span!
+                       {:name "instaql/rule-where-exception"
+                        :attributes {:code (:code (rule-model/get-program! rules
+                                                                           etype
+                                                                           "view"))
+                                     :error e}}
+                     acc))))
+             {}
+             program-results)]
+        (tracer/add-data! {:attributes {:rule-wheres rule-wheres}})
+        rule-wheres))))
+
+(defn update-where-with-rule-refs [initial-etype attrs rule-wheres where k v]
+  (loop [[segment & rest-path] (string/split (name k) #"\.")
+         current-path [segment]
+         etype initial-etype
+         where (assoc where k v)]
+    (if-let [segment-etype (link-etype attrs etype segment)]
+      (let [rule-where (if (-> rule-wheres (get segment-etype) :short-circuit?)
+                         {:id (str (random-uuid))}
+                         (-> rule-wheres (get segment-etype) :where-clauses))
+            where (if-not (seq rule-where)
+                    where
+                    (update where
+                            :and
+                            (fnil conj [])
+                            (prefix-all-clauses (string/join "." current-path) rule-where)))]
+        (if-not (seq rest-path)
+          where
+          (recur rest-path
+                 (conj current-path (first rest-path))
+                 segment-etype
+                 where)))
+      where)))
+
+(defn extend-where-with-rule-refs [etype ctx rule-wheres where]
+  (reduce-kv (fn [new-where k v]
+               (if (or (or-where-cond? [k v])
+                       (and-where-cond? [k v]))
+                 (assoc new-where k (concat (map (partial extend-where-with-rule-refs
+                                                          ctx
+                                                          rule-wheres)
+                                                 v)
+                                            ;; We may have added our own ands/or
+                                            ;; in add-rule-wheres-to-query
+                                            (get new-where k)))
+                 (update-where-with-rule-refs etype
+                                              (:attrs ctx)
+                                              rule-wheres
+                                              new-where
+                                              k
+                                              v)))
+             {}
+             where))
+
+(defn add-rule-wheres-to-query
+  ([ctx rule-wheres o]
+   (add-rule-wheres-to-query nil ctx rule-wheres o))
+  ([etype ctx rule-wheres o]
+   (reduce-kv (fn [new-o k v]
+                (let [etype (if etype
```

Comment:
Wanted to share a chain of thought: 

For self-referencing relationships, i.e : blocks have children blocks. 

When a query is run like: 

```js
{ blocks: { children: { } } }
```

How would the where clauses attach? 

I noticed we use the `etype` to get the where clauses. So say it had something like: 

rule: !data.parent || data.parent == "some-uuid" 

The enhanced query would be: 

```
{ blocks: { 
  $: { where: { or: [{parent: null}, parent: "some-uuid"] } }
  children: {
       $: { where: { or: [{parent: null}, parent: "some-uuid"] } }
   } 
} }
```

Which, though surprising, I _think_ is what it should it do! Cool. 

--- 

Do I understand this correctly?